src/java/org/apache/cassandra/index/sai/disk/v1/vector/OnHeapGraph.java

```diff
@@ -0,0 +1,360 @@
+     * @param concurrent should be true for memtables, false for compaction.  Concurrent allows us to search
+     *                   while building the graph; non-concurrent allows us to avoid synchronization costs.
+     */
+    @SuppressWarnings("unchecked")
+    public OnHeapGraph(AbstractType<?> termComparator, IndexWriterConfig indexWriterConfig, boolean concurrent)
+    {
+        this.vectorType = (VectorType<?>) termComparator;
+        vectorValues = concurrent
+                       ? new ConcurrentVectorValues(((VectorType<?>) termComparator).dimension)
+                       : new CompactionVectorValues(((VectorType<Float>) termComparator));
+        similarityFunction = indexWriterConfig.getSimilarityFunction();
+        // We need to be able to inexpensively distinguish different vectors, with a slower path
+        // that identifies vectors that are equal but not the same reference.  A comparison
+        // based Map (which only needs to look at vector elements until a difference is found)
+        // is thus a better option than hash-based (which has to look at all elements to compute the hash).
+        postingsMap = new ConcurrentSkipListMap<>(Arrays::compare);
+        postingsByOrdinal = new NonBlockingHashMapLong<>();
+
+        builder = new GraphIndexBuilder<>(vectorValues,
+                                          VectorEncoding.FLOAT32,
+                                          similarityFunction,
+                                          indexWriterConfig.getMaximumNodeConnections(),
+                                          indexWriterConfig.getConstructionBeamWidth(),
+                                          1.2f,
+                                          1.4f);
+    }
+
+    public int size()
+    {
+        return vectorValues.size();
+    }
+
+    public boolean isEmpty()
+    {
+        return postingsMap.values().stream().allMatch(VectorPostings::isEmpty);
+    }
+
+    /**
+     * @return the incremental bytes ysed by adding the given vector to the index
+     */
+    public long add(ByteBuffer term, T key, InvalidVectorBehavior behavior)
+    {
+        assert term != null && term.remaining() != 0;
+
+        var vector = vectorType.composeAsFloat(term);
+        if (behavior == InvalidVectorBehavior.IGNORE)
+        {
+            try
+            {
+                validateIndexable(vector, similarityFunction);
+            }
+            catch (InvalidRequestException e)
+            {
+                logger.trace("Ignoring invalid vector during index build against existing data: {}", vector, e);
+                return 0;
+            }
+        }
+        else
+        {
+            assert behavior == InvalidVectorBehavior.FAIL;
+            validateIndexable(vector, similarityFunction);
+        }
+
+        var bytesUsed = new AtomicLong();
+        var newVector = new AtomicBoolean();
+        // if the vector is already in the graph, all that happens is that the postings list is updated
```

Comment:
this has a pretty nasty race condition when multiple rows are added concurrently with the same vector.  fix is in vsearch 9e3ca22c